mm2src/mm2_main/src/lp_swap/swap_watcher.rs

```diff
@@ -0,0 +1,466 @@
+
+        Ok((Some(WatcherCommand::Finish), vec![WatcherEvent::MakerPaymentSpent(
+            tx_ident,
+        )]))
+    }
+}
+
+pub async fn run_watcher(swap: RunWatcherInput, ctx: MmArc) {
+    let swap_ctx = SwapsContext::from_ctx(&ctx).unwrap();
+    let uuid = swap.uuid().to_owned();
+
+    let (swap, mut command) = match swap {
+        RunWatcherInput::StartNew(swap) => (swap, WatcherCommand::Start),
+    };
+
+    let ctx = swap.ctx.clone();
+    let mut status = ctx.log.status_handle();
+    let uuid_str = swap.uuid.to_string();
+    let running_swap = Arc::new(swap);
+    let shutdown_rx = swap_ctx.shutdown_rx.clone();
+    let swap_for_log = running_swap.clone();
+
+    let mut swap_fut = Box::pin(
+        async move {
+            let mut events;
+            loop {
+                let res = running_swap.handle_command(command).await.expect("!handle_command");
+                events = res.1;
+                for event in events {
+                    status.status(&[&"swap", &("uuid", uuid_str.as_str())], &event.status_str());
+                    running_swap.apply_event(event);
+                }
+                match res.0 {
+                    Some(c) => {
+                        command = c;
+                    },
+                    None => {
+                        break;
+                    },
+                }
+            }
+        }
+        .fuse(),
+    );
+    let mut shutdown_fut = Box::pin(shutdown_rx.recv().fuse());
+    let do_nothing = (); // to fix https://rust-lang.github.io/rust-clippy/master/index.html#unused_unit
+    select! {
+        _swap = swap_fut => {
+            swap_ctx.taker_swap_watchers.lock().await.remove(&uuid);
+            do_nothing
+        }, // swap finished normally
+        _shutdown = shutdown_fut => {
+            swap_ctx.taker_swap_watchers.lock().await.remove(&uuid);
+            info!("swap {} stopped!", swap_for_log.uuid)
+        },
+    };
+}
+
+pub async fn process_watcher_msg(ctx: MmArc, msg: &[u8]) {
+    let msg = match decode_signed::<SwapWatcherMsg>(msg) {
+        Ok(m) => m,
+        Err(watcher_msg_err) => {
+            error!("Couldn't deserialize 'SwapWatcherMsg': {:?}", watcher_msg_err);
+            // Drop it to avoid dead_code warning
+            drop(watcher_msg_err);
+            return;
+        },
+    };
+
+    match msg.0 {
+        SwapWatcherMsg::TakerSwapWatcherMsg(watcher_data) => spawn_taker_swap_watcher(ctx, *watcher_data),
+    }
+}
+
+fn spawn_taker_swap_watcher(ctx: MmArc, watcher_data: TakerSwapWatcherData) {
+    let swap_ctx = SwapsContext::from_ctx(&ctx).unwrap();
+    let mut taker_swap_watchers = match swap_ctx.taker_swap_watchers.try_lock() {
+        Some(lock) => lock,
+        None => return,
+    };
+    // Return if taker/maker swap or watcher with the same uuid already exists
+    if swap_ctx.swap_msgs.lock().unwrap().contains_key(&watcher_data.uuid)
+        || taker_swap_watchers.contains(&watcher_data.uuid)
```

Comment:
Could you please check if `swap_ctx.swap_msgs.lock().unwrap().contains_key(&watcher_data.uuid)` before `taker_swap_watchers` is locked?
It will allow us to avoid deadlocks in the future